src/frontend/src/optimizer/plan_node/stream_hash_join.rs

```diff
@@ -82,7 +94,70 @@ impl StreamHashJoin {
                     }
                 }
             }
-            logical.i2o_col_mapping().rewrite_bitset(&watermark_columns)
+            let (left_cols_num, original_inequality_pairs) = eq_join_predicate.inequality_pairs();
+            let mut band_condition_clean_left_state = false;
+            let mut band_condition_clean_right_state = false;
+            for (
+                conjunction_idx,
+                InequalityInputPair {
+                    key_required_larger,
+                    key_required_smaller,
+                    delta_expression,
+                },
+            ) in original_inequality_pairs
+            {
+                let both_upstream_has_watermark = if key_required_larger < key_required_smaller {
+                    logical
+                        .left()
+                        .watermark_columns()
+                        .contains(key_required_larger)
+                        && logical
+                            .right()
+                            .watermark_columns()
+                            .contains(key_required_smaller - left_cols_num)
+                } else {
+                    logical
+                        .left()
+                        .watermark_columns()
+                        .contains(key_required_smaller)
+                        && logical
+                            .right()
+                            .watermark_columns()
+                            .contains(key_required_larger - left_cols_num)
+                };
+                if !both_upstream_has_watermark {
+                    continue;
+                }
+
+                clean_state_conjunction_indices.push(conjunction_idx);
+                let internal = if key_required_larger < key_required_smaller {
+                    band_condition_clean_left_state = true;
+                    l2i.try_map(key_required_larger)
+                } else {
+                    band_condition_clean_right_state = true;
+                    r2i.try_map(key_required_larger - left_cols_num)
+                };
+                let generate_watermark = if let Some(internal) = internal && !watermark_columns.contains(internal) {
+                    watermark_columns.insert(internal);
+                    true
+                } else {
+                    false
+                };
```

Comment:
Can we just keep those band conditions with generate_watermark = true? If they don't generate watermarks, then we have no reason to keep them.

Reply:
> Can we just keep those band conditions with generate_watermark = true? If they don't generate watermarks, then we have no reason to keep them.

Although the column mentioned in those conditions are not included in executor output so the watermark is not visible in the view of downstream, we can use the purely internal watermark to clean state.

Reply:
> Can we just keep those band conditions with generate_watermark = true? If they don't generate watermarks, then we have no reason to keep them.

ok I have just deleted `generate_watermark = false` conditions.